rfcs/0127-issues-warnings.md

```diff
@@ -0,0 +1,283 @@
+
+For some use cases, like for packages without maintainers, we do not want to break evaluation of a package and simply warn the user instead. We want the users to configure all these according to their needs and through a single standardized interface.
+
+## Detailed design
+[design]: #detailed-design
+
+### Package problems
+
+A new attribute is added to the `meta` section of a package: `problems`. If present, it is a list of attrsets which each have at least the following fields:
+
+- `kind`: Required. If present, the resulting warning will be printed as `kind: message`.
+- `message`: Required. A string message describing the issue with the package. The value should:
+  - Start with the "This package", "The application" or equivalent, or simply with the package name.
+  - Be capitalized (unless it starts with the package name).
+  - Use a period at the end.
+- `name`: Required if there are multiple values of the same `kind`. Give the issue a custom name for more easy filtering
+- `date`: Required. An ISO 8601 `yyyy-mm-dd`-formatted date from when the issue was added.
+- `urls`: Optional, list of strings. Can be used to link issues, pull requests and other related items.
+
+Other attributes are allowed. Some message kinds may specify additional required attributes.
+
+Example values:
+
+```nix
+meta.problems = [
+  {
+    name = "python2-eol";
+    kind = "deprecated";
+    message = "This package depends on Python 2, which has reached end of life.";
+    date = "1970-01-01";
+    urls = [ "https://github.com/NixOS/nixpkgs/issues/148779" ];
+  }
+  {
+    kind = "removal";
+    message = "The application has been abandoned upstream, use libfoo instead";
+    date = "1970-01-01";
+  }
+];
+```
+
+### Problem kinds
+
+At the moment, the following values for the `kind` field of a warning are known:
+
+- `removal`: The package is scheduled for removal some time in the future.
+- `deprecated`: The package has been abandoned upstream or has end of life dependencies.
+- `maintainerless`: `meta.maintainers` is empty
+- `insecure`: The package has some security vulnerabilities
+- `broken`: The package is marked as broken
+- `unsupported`: The package is not expected to build on this platform
+
+Not all values make sense for declaration in `meta.problems`: Some may be automatically generated from other `meta` attributes (for example `maintainerless`). New kinds may be added in the future. Furthermore, some kinds are expected to be present only up to once per derivation: for example, we have no use for having multiple `maintainerless` problems, and therefore also no need to give them a name in order to distinguish them.
```

Comment:
Would anything be generated from `sourceProvenance = lib.sourceTypes.binaryNativeCode;`? I'm of the opinion that not having the option to turn off substitution and build from source makes a package unsecure, but then some only consider that to be an inconvenience (because it's not patchable), so maybe a separate category? And you'd probably want to not flag bootstrapping seeds with a problem that can never be resolved.

Or maybe I should be saving these questions until after the proposal is accepted.

Reply:
Yes, to which extent the source provenance thing can be integrated into this RFC is an open question, and while we can already discuss it now I wouldn't commit on any of that in the RFC itself (apart from maybe mentioning it as potential future work).

Reply:
> I'm of the opinion that not having the option to turn off substitution and build from source makes a package unsecure,

Me too, so I set `allowNonSource = false;` in `nixpkgs.nix`.  Is there something you want that isn't achieved by that?

Reply:
No, I was just wondering if that had been considered.